_posts/2023-04-12-making-verification-compelling-visual-verification-feedback-for-dafny.markdown

```diff
@@ -0,0 +1,253 @@
+
+Gutter icons brought the first solution in that they offered easy-to-read and enhanced contextual feedback.
+You might remember the animated screenshot of [our previous blog post](/blog/2023/03/03/dafny-4-released/):
+
+![Dafny 4 IDE features](/blog/assets/images/Dafny4IDEFeatures.gif)
+
+I'm now going to explain in detail what is happening on the left of the code (the _gutter_), and how I designed it and especially how it answered many of the questions users were asking, sometimes unconciously.
+
+Numerous test engines also display feedback in the gutter next to the methods themselves, or next to the method calls if they are tests (e.g. [wallabyjs](https://wallabyjs.com/) [xUnit](https://xunit.net/docs/getting-started/netfx/jetbrains-rider)...)}.
+Such feedback is contextual, meaning it can be associated to the line of code it is attached to, as opposed to diagnostics that are listed in a separate window.
+Because unit tests are much less common than pre- and postconditions in Dafny,
+and because I had the possibility to customize the icon for every line, I wondered:
+how can I leverage the gutter area to display verification feedback instead of usual test feedback? 
+
+While thinking about how to display verification feedback in the gutter, I established some principles that guided us during the design of both the icons and the user experience:
+
+* **Verified is visible but the least distracting:** The icons for verified code should provide a simple positive feedback.
+* **Accessibility:** The icons should not just have colors, but also easily recognizable shapes, so that color-blind people could still benefit from them.
+* **Compatibility with themes:** The icon colors should be adequate both in light and dark mode.
+* **Compatibility with modes:** The icons should work both if the user triggers verification on save, or if verification is triggered each time users type something
+* **Context-awareness:** The icons should not just convey the verification status of a single line, but, where appropriate, they may also convey verification feedback about the surrounding context.
+* **Past-awareness:** Icons indicating ongoing computation should still display previous verification status, so that it is not lost.
+* **No ambiguity:** The icons should be simple, straightforward, and not be ambiguous when looked from the side of the field of vision, so that users can focus on their code.
+
+See below these icons in the current version of Dafny, on Visual Studio Code, when other gutter icon features are disabled.
+
+<div id="fig_static_icons_dark_light"></div>
+[<img src="/blog/assets/images/verification-compelling/example1.png" alt="Dafny gutter icons on dark theme" width="45%"/>](/blog/assets/images/verification-compelling/example1.png) [<img src="/blog/assets/images/verification-compelling/example2.png" alt="Dafny gutter icons on light theme" width="45%"/>](/blog/assets/images/verification-compelling/example2.png)
```

Comment:
This turns the images into clickable links. Is that your intention?

Reply:
Yes it's my intention.